### PR TITLE
Implement error types for functions

### LINKED ISSUE #356: Function errors not captured
When a function errors for some reason (uncaught exception), dispatch does not provide any helpful output, only a cryptic serialization error.

- [x] Bug
- [ ] Feature
- [ ] Enhancement

## Detailed Description

The following function throws an exception:
```
def handle(ctx, payload):
    raise Exception("oh no!")
```
When run the output does not point to the issue:
```
$ dispatch exec error --wait
{
    "blocking": true,
    "executedTime": 1522974627,
    "finishedTime": -62135596800,
    "functionId": "75f6a73a-e2d4-40a1-9e4e-fa99add9279c",
    "functionName": "error",
    "input": {},
    "logs": null,
    "name": "436f40fe-7e47-4be8-9711-237fe39ee0cf",
    "reason": [
        "error running function: error: cannot JSON-parse result from OpenFaaS: invalid character 'I' looking for beginning of value Internal Server Error\r\n\r\nThe server encountered an unexpected internal server error\r\n\r\n(generated by waitress)"
    ],
    "secrets": [],
    "services": null,
    "status": "ERROR",
    "tags": []
}
```
I would expect a traceback to be captured somewhere in the run output which would help tremendously for debugging.

## Complexity

- [ ] Low - Simple enhancement or bug fix, no architectural changes or refactoring
- [x] Medium - Change requires some thought, but is relatively isolated
- [ ] High - Significant architectural change or large refactor

## Impact

- [ ] Low - Annoyance, but does not impact business or functionality
- [x] Medium - Issue can be worked around, but is causing pain
- [ ] High - Blocker

## Your Environment

* Dispatch CLI version (or git commit):
* Dispatch Chart version (or image[s] tag):
* Operating System and version:
* Kubernetes version (and distribution):
* Etc (any other useful environmental information):

### LINKED ISSUE #403: Function error not displayed
- [x] Bug
- [ ] Feature
- [ ] Enhancement

## Detailed Description

When an error occurs when running a function, the error is returned as part of the context but we don't display this to the user. It would be nice to display the error that occurred and set the function run status as "ERROR".

Currently when a function throws an exception,
```
def handle(ctx, payload):
    raise Exception("oh no!")
```

A traceback is returned in the stderr, but the function still reports as "READY" and reason is `null`
```
$ dispatch exec fail --wait
{
    "blocking": true,
    "executedTime": 1524863021,
    "faasId": "00bbecbe-9815-44ce-8445-56715a107c6f",
    "finishedTime": 1524863021,
    "functionId": "65f708fd-61b9-4949-af1d-fba822d3abea",
    "functionName": "fail",
    "input": {},
    "logs": {
        "stderr": [
            "Traceback (most recent call last):",
            "  File \"main.py\", line 43, in exec_function",
            "    r = handler.handle(msg['context'], msg['payload'])",
            "  File \"/function-server/function/handler.py\", line 2, in handle",
            "    raise Exception(\"oh no!\")",
            "Exception: oh no!"
        ],
        "stdout": null
    },
    "name": "19f2a8ae-6010-4ad5-98f5-5cd611f09101",
    "reason": null,
    "secrets": [],
    "services": null,
    "status": "READY",
    "tags": []
}
```

## Complexity

- [x] Low - Simple enhancement or bug fix, no architectural changes or refactoring
- [ ] Medium - Change requires some thought, but is relatively isolated
- [ ] High - Significant architectural change or large refactor

## Impact

- [x] Low - Annoyance, but does not impact business or functionality
- [ ] Medium - Issue can be worked around, but is causing pain
- [ ] High - Blocker

## Your Environment

* Dispatch CLI version (or git commit):
* Dispatch Chart version (or image[s] tag):
* Operating System and version:
* Kubernetes version (and distribution):
* Etc (any other useful environmental information):

### PR DESCRIPTION
**Pull request by rjew**
Fixes #356, Fixes #403

* Implement error handling spec
* Distinguish between error types: InputError, FunctionError, SystemError
* Displays the error and its properties (type, stacktrace, etc.) when a function invocation encounters an error
* Added unit tests and a small e2e test
* Requires some changes in the language packs

### CODE REVIEW

File: pkg/functions/openfaas/driver.go [hunk 1]
```diff
@@ -227,21 +251,22 @@ func (d *ofDriver) GetRunnable(e *functions.FunctionExecution) functions.Runnabl
 		case 200:
 			resBytes, err := ioutil.ReadAll(res.Body)
 			if err != nil {
-				return nil, errors.Errorf("cannot read result from OpenFaaS on URL: %s %s", d.gateway, err)
+				return nil, &systemError{errors.Errorf("cannot read result from OpenFaaS on URL: %s %s", d.gateway, err)}
 			}
 			var out functions.Message
 			if err := json.Unmarshal(resBytes, &out); err != nil {
-				return nil, errors.Errorf("cannot JSON-parse result from OpenFaaS: %s %s", err, string(resBytes))
+				return nil, &outputError{errors.Errorf("cannot JSON-parse result from OpenFaaS: %s %s", err, string(resBytes))}
```
Comment (imikushin) [inline pkg/functions/openfaas/driver.go]: This one is a system error too: it's the function runtime responsible for JSON-encoding the response, not the function.

File: pkg/functions/riff/driver.go [hunk 2]
```diff
@@ -116,14 +140,15 @@ func (d *riffDriver) GetRunnable(e *functions.FunctionExecution) functions.Runna
...
 		resBytes, err := d.requester.Request(topic, e.RunID, bytesIn)
 		if err != nil {
-			return nil, errors.Wrapf(err, "riff: error invoking function: '%s', runID: '%s'", e.FunctionID, e.RunID)
+			return nil, &systemError{errors.Wrapf(err, "riff: error invoking function: '%s', runID: '%s'", e.FunctionID, e.RunID)}
 		}
 
 		var out functions.Message
 		if err := json.Unmarshal(resBytes, &out); err != nil {
-			return nil, errors.Errorf("cannot JSON-parse result from riff: %s %s", err, string(resBytes))
+			return nil, &outputError{errors.Errorf("cannot JSON-parse result from riff: %s %s", err, string(resBytes))}
```
Comment (imikushin) [inline pkg/functions/riff/driver.go]: This one is a system error too: it's the function runtime responsible for JSON-encoding the response, not the function.

**Comment by berndtj** [comment]
@rjew can you rebase after the big model refactor

**Comment by rjew** [comment]
Sure working on it now

**Pull request merged**